crates/key-derivation/src/operator.rs

```diff
@@ -0,0 +1,276 @@
+            ChildNumber::from_hardened_idx(OPERATOR_IDX).unwrap(),
+            ChildNumber::from_hardened_idx(WALLET_IDX).unwrap(),
+        ]);
+
+        let message_xpriv = master.derive_priv(SECP256K1, &message_path)?;
+        let wallet_xpriv = master.derive_priv(SECP256K1, &wallet_path)?;
+
+        Ok(Self {
+            master: *master,
+            signing: message_xpriv,
+            wallet: wallet_xpriv,
+        })
+    }
+
+    /// Operator's master [`Xpriv`].
+    pub fn master_xpriv(&self) -> Xpriv {
+        self.master
+    }
+
+    /// Operator's wallet transaction signing [`Xpriv`].
+    pub fn wallet_xpriv(&self) -> Xpriv {
+        self.wallet
+    }
+
+    /// Operator's message signing [`Xpriv`].
+    pub fn message_xpriv(&self) -> Xpriv {
+        self.signing
+    }
+
+    /// Operator's master [`Xpub`].
+    pub fn master_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.master)
+    }
+
+    /// Operator's message signing [`Xpub`].
+    ///
+    /// Infallible according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    pub fn message_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.signing)
+    }
+
+    /// Operator's wallet transaction signing [`Xpub`].
+    ///
+    /// Infallible conversion from [`Xpriv`] to [`Xpub`] according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    pub fn wallet_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.wallet)
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl Zeroize for OperatorKeys {
+    fn zeroize(&mut self) {
+        let Self {
+            master,
+            signing,
+            wallet,
+        } = self;
+
+        // # Security note
+        //
+        // Going over all possible "zeroizable" fields.
+        // What we cannot zeroize is:
+        //
+        // - Network
+        // - Child number
+        //
+        // These are fine to leave as they are since they are public parameters,
+        // and not secret values.
+
+        master.depth.zeroize();
+        let mut parent_fingerprint: [u8; 4] = *master.parent_fingerprint.as_mut();
+        parent_fingerprint.zeroize();
+        master.private_key.non_secure_erase();
+        let mut chaincode: [u8; 32] = *master.chain_code.as_mut();
+        chaincode.zeroize();
+
+        signing.depth.zeroize();
+        let mut parent_fingerprint: [u8; 4] = *signing.parent_fingerprint.as_mut();
+        parent_fingerprint.zeroize();
+        signing.private_key.non_secure_erase();
+        let mut chaincode: [u8; 32] = *signing.chain_code.as_mut();
+        chaincode.zeroize();
+
+        wallet.depth.zeroize();
+        let mut parent_fingerprint: [u8; 4] = *wallet.parent_fingerprint.as_mut();
+        parent_fingerprint.zeroize();
+        wallet.private_key.non_secure_erase();
+        let mut chaincode: [u8; 32] = *wallet.chain_code.as_mut();
+        chaincode.zeroize();
```

Comment:
Doesn't `Xpriv` support its own internal zeroization via a custom method that takes care of the entire struct? The technique used here relies heavily on knowledge of its internals.

Reply:
No it does not. At least is not exposed, see https://docs.rs/bitcoin/latest/bitcoin/bip32/struct.Xpriv.html

But the fields are public and that's what I did. Follow the fields and the ones that I was able to `zeroize`, I did.

Reply:
Got it! I misunderstood the scope of this `Xpriv` functionality. Unfortunate that they don't just support `Zeroize` directly...